source/adapters/cuda/enqueue.cpp

```diff
@@ -561,17 +591,21 @@ UR_APIEXPORT ur_result_t UR_APICALL urEnqueueKernelLaunchCustomExp(
   size_t ThreadsPerBlock[3] = {32u, 1u, 1u};
   size_t BlocksPerGrid[3] = {1u, 1u, 1u};
 
+  // Set work group memory so we can compute the whole memory requirement
+  if (WorkGroupMemory)
+    hKernel->setWorkGroupMemory(WorkGroupMemory);
   uint32_t LocalSize = hKernel->getLocalSize();
   CUfunction CuFunc = hKernel->get();
 
   for (uint32_t i = 0; i < numPropsInLaunchPropList; i++) {
     switch (launchPropList[i].id) {
     case UR_EXP_LAUNCH_PROPERTY_ID_IGNORE: {
+      launch_attribute.push_back({});
       launch_attribute[i].id = CU_LAUNCH_ATTRIBUTE_IGNORE;
```

Comment:
If we hit the `UR_EXP_LAUNCH_PROPERTY_ID_WORK_GROUP_MEMORY` case, we don't push back a `CUlaunchAttribute`. At that point, the value of `i` goes out of sync with the number of elements in `launch_attribute`, and `launch_attribute[i]` is out of bounds. Maybe this is fine because `UR_EXP_LAUNCH_PROPERTY_ID_WORK_GROUP_MEMORY` only ever comes after the others?

I'd suggest we remove the reliance on `launch_attribute[i]` and use `launch_attribute.back()`.

Since we're C++17, we could do `auto &attr = launch_attribute.emplace_back({}); attr.id = ...`, right?

Reply:
good catch, fixed. 

> Since we're C++17, we could do auto &attr = launch_attribute.emplace_back({}); attr.id = ..., right?

why not, cleaner than calling `launch_attribute.back()` over and over